File: TopInterview150/238_product_of_array_except_self/productExceptSelf.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        n = len(nums)
        pre = 1
        post = 1

        for i in range(n):
            res[i] *= pre
            pre *= nums[i]
            res[n - i - 1] *= post
            post *= nums[n - i - 1]

        return res

File: TopInterview150/238_product_of_array_except_self/productExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        int prefix = 1;
        int postfix = 1;

        Arrays.fill(result, 1);

        for (int i = 0; i < n; i++) {
            result[i] *= prefix;
            prefix *= nums[i];

            result[n - i - 1] *= postfix;
            postfix *= nums[n - i - 1];
        }

        return result;
    }
}